iroh/src/downloader.rs

```diff
@@ -0,0 +1,979 @@
+        let as_blob = DownloadKind::Blob { hash };
+        let as_collection = DownloadKind::Collection { hash };
+        self.current_requests.contains_key(&as_blob)
+            || self.scheduled_requests.contains_key(&as_blob)
+            || self.current_requests.contains_key(&as_collection)
+            || self.scheduled_requests.contains_key(&as_collection)
+    }
+
+    /// Handle receiving a new connection.
+    fn on_connection_ready(&mut self, peer: PublicKey, result: anyhow::Result<D::Connection>) {
+        match result {
+            Ok(connection) => {
+                trace!(%peer, "connected to peer");
+                let drop_key = self.goodbye_peer_queue.insert(peer, IDLE_PEER_TIMEOUT);
+                self.peers
+                    .insert(peer, ConnectionInfo::new_idle(connection, drop_key));
+            }
+            Err(err) => {
+                debug!(%peer, %err, "connection to peer failed")
+            }
+        }
+    }
+
+    fn on_download_completed(&mut self, kind: DownloadKind, result: Result<(), FailureAction>) {
+        // first remove the request
+        let info = self
+            .current_requests
+            .remove(&kind)
+            .expect("request was active");
+
+        // update the active requests for this peer
+        let ActiveRequestInfo {
+            intents,
+            peer,
+            mut remaining_retries,
+            ..
+        } = info;
+
+        let peer_info = self.peers.get_mut(&peer).expect("peer is connected");
+        peer_info.state = match &peer_info.state {
+            PeerState::Busy { active_requests } => {
+                match NonZeroUsize::new(active_requests.get() - 1) {
+                    Some(active_requests) => PeerState::Busy { active_requests },
+                    None => {
+                        // last request of the peer was this one
+                        let drop_key = self.goodbye_peer_queue.insert(peer, IDLE_PEER_TIMEOUT);
+                        PeerState::Idle { drop_key }
+                    }
+                }
+            }
+            PeerState::Idle { .. } => unreachable!("peer was busy"),
+        };
+
+        let hash = *kind.hash();
+
+        match result {
+            Ok(()) => {
+                debug!(%peer, ?kind, "download completed");
+                for sender in intents.into_values() {
+                    let _ = sender.send(Ok(()));
+                }
+            }
+            Err(FailureAction::AbortRequest(reason)) => {
+                debug!(%peer, ?kind, %reason, "aborting request");
+                for sender in intents.into_values() {
+                    let _ = sender.send(Err(()));
+                }
+            }
+            Err(FailureAction::DropPeer(reason)) => {
+                debug!(%peer, ?kind, %reason, "peer will be dropped");
+                if let Some(_connection) = peer_info.conn.take() {
+                    // TODO(@divma): this will fail open streams, do we want this?
```

Comment:
lets at least log this in a way that we can see if this is an issue

Reply:
we log it in the previous line. Something else you think we should log or that should be added?

Reply:
Maybe the connection id?